pkg/services/secrets/kvstore/migrations/datasources_mig.go

```diff
@@ -0,0 +1,85 @@
+package migrations
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/grafana/grafana/pkg/bus"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/datasources"
+	"github.com/grafana/grafana/pkg/services/featuremgmt"
+	"github.com/grafana/grafana/pkg/services/secrets/kvstore"
+	"github.com/grafana/grafana/pkg/services/sqlstore"
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+const (
+	dataSourceSecretType = "datasource"
+)
+
+type DataSourceSecretMigrationService struct {
+	sqlStore           *sqlstore.SQLStore
+	dataSourcesService datasources.DataSourceService
+	secretsStore       kvstore.SecretsKVStore
+	features           featuremgmt.FeatureToggles
+	log                log.Logger
+	bus                bus.Bus
+}
+
+func ProvideDataSourceMigrationService(
+	cfg *setting.Cfg, dataSourcesService datasources.DataSourceService,
+	secretsStore kvstore.SecretsKVStore, features featuremgmt.FeatureToggles,
+	sqlStore *sqlstore.SQLStore, bus bus.Bus,
+) *DataSourceSecretMigrationService {
+	return &DataSourceSecretMigrationService{
+		sqlStore:           sqlStore,
+		dataSourcesService: dataSourcesService,
+		secretsStore:       secretsStore,
+		features:           features,
+		log:                logger,
+		bus:                bus,
+	}
+}
+
+func (s *DataSourceSecretMigrationService) Run(ctx context.Context) error {
+	return s.sqlStore.InTransaction(ctx, func(ctx context.Context) error {
+		query := &datasources.GetDataSourcesQuery{}
+		err := s.dataSourcesService.GetDataSources(ctx, query)
+		if err != nil {
+			return err
+		}
+		for _, ds := range query.Result {
+			hasMigration, _ := ds.JsonData.Get("secretMigrationComplete").Bool()
+			if !hasMigration {
+				secureJsonData, err := s.dataSourcesService.DecryptLegacySecrets(ctx, ds)
+				if err != nil {
+					return err
+				}
+
+				jsonData, err := json.Marshal(secureJsonData)
+				if err != nil {
+					return err
+				}
+
+				err = s.secretsStore.Set(ctx, ds.OrgId, ds.Name, dataSourceSecretType, string(jsonData))
+				if err != nil {
+					return err
+				}
+
+				ds.JsonData.Set("secretMigrationComplete", true)
+				err = s.dataSourcesService.UpdateDataSource(ctx, &datasources.UpdateDataSourceCommand{Id: ds.Id, OrgId: ds.OrgId, Uid: ds.Uid, JsonData: ds.JsonData})
+				if err != nil {
+					return err
+				}
+			}
+
+			if s.features.IsEnabled(featuremgmt.FlagDisableSecretsCompatibility) && len(ds.SecureJsonData) > 0 {
+				err := s.dataSourcesService.DeleteDataSourceSecrets(ctx, &datasources.DeleteDataSourceSecretsCommand{UID: ds.Uid, OrgID: ds.OrgId, ID: ds.Id})
```

Comment:
Maybe introducing a new field on the UpdateDataSourceCommand telling if secrets should be cleaned or not might be an alternative to introducing this method.

Reply:
Actually, I believe we can already do that with the update data source command by setting the secrets to an empty map. Will update the implementation so this method is not needed.

Reply:
The field on UpdateDataSourceCommand wasn't necessary because that's already controled by the disableSecretsCompatibility flag.

Reply:
The UpdateDataSource function was indeed already taking care of using the correct database and deleting secrets from legacy databases based on the compatibility flag, so the DeleteDataSourceSecrets function wasn't necessary.

Reply:
Removed on https://github.com/grafana/grafana/pull/50463/commits/bb1af49b905deceeff489b78f68247fd39e82977